File: JeMPI_Apps/JeMPI_LibMPI/src/main/java/org/jembi/jempi/libmpi/dgraph/CustomDgraphConstants.java
package org.jembi.jempi.libmpi.dgraph;

public final class CustomDgraphConstants {

   public static final String PREDICATE_GOLDEN_RECORD_AUX_DATE_CREATED = "GoldenRecord.aux_date_created";
   public static final String PREDICATE_GOLDEN_RECORD_AUX_AUTO_UPDATE_ENABLED = "GoldenRecord.aux_auto_update_enabled";
   public static final String PREDICATE_GOLDEN_RECORD_AUX_ID = "GoldenRecord.aux_id";
   public static final String PREDICATE_GOLDEN_RECORD_GIVEN_NAME = "GoldenRecord.given_name";
   public static final String PREDICATE_GOLDEN_RECORD_FAMILY_NAME = "GoldenRecord.family_name";
   public static final String PREDICATE_GOLDEN_RECORD_GENDER = "GoldenRecord.gender";
   public static final String PREDICATE_GOLDEN_RECORD_DOB = "GoldenRecord.dob";
   public static final String PREDICATE_GOLDEN_RECORD_CITY = "GoldenRecord.city";
   public static final String PREDICATE_GOLDEN_RECORD_PHONE_NUMBER = "GoldenRecord.phone_number";
   public static final String PREDICATE_GOLDEN_RECORD_NATIONAL_ID = "GoldenRecord.national_id";
   public static final String PREDICATE_GOLDEN_RECORD_INTERACTIONS = "GoldenRecord.interactions";
   public static final String PREDICATE_INTERACTION_AUX_DATE_CREATED = "Interaction.aux_date_created";
   public static final String PREDICATE_INTERACTION_AUX_ID = "Interaction.aux_id";
   public static final String PREDICATE_INTERACTION_AUX_CLINICAL_DATA = "Interaction.aux_clinical_data";
   public static final String PREDICATE_INTERACTION_GIVEN_NAME = "Interaction.given_name";
   public static final String PREDICATE_INTERACTION_FAMILY_NAME = "Interaction.family_name";
   public static final String PREDICATE_INTERACTION_GENDER = "Interaction.gender";
   public static final String PREDICATE_INTERACTION_DOB = "Interaction.dob";
   public static final String PREDICATE_INTERACTION_CITY = "Interaction.city";
   public static final String PREDICATE_INTERACTION_PHONE_NUMBER = "Interaction.phone_number";
   public static final String PREDICATE_INTERACTION_NATIONAL_ID = "Interaction.national_id";

   static final String GOLDEN_RECORD_FIELD_NAMES =
         """
         uid
         GoldenRecord.source_id {
            uid
            SourceId.facility
            SourceId.patient
         }
         GoldenRecord.aux_date_created
         GoldenRecord.aux_auto_update_enabled
         GoldenRecord.aux_id
         GoldenRecord.given_name
         GoldenRecord.family_name
         GoldenRecord.gender
         GoldenRecord.dob
         GoldenRecord.city
         GoldenRecord.phone_number
         GoldenRecord.national_id
         """;

   static final String EXPANDED_GOLDEN_RECORD_FIELD_NAMES =
         """
         uid
         GoldenRecord.source_id {
            uid
            SourceId.facility
            SourceId.patient
         }
         GoldenRecord.aux_date_created
         GoldenRecord.aux_auto_update_enabled
         GoldenRecord.aux_id
         GoldenRecord.given_name
         GoldenRecord.family_name
         GoldenRecord.gender
         GoldenRecord.dob
         GoldenRecord.city
         GoldenRecord.phone_number
         GoldenRecord.national_id
         GoldenRecord.interactions @facets(score) {
            uid
            Interaction.source_id {
               uid
               SourceId.facility
               SourceId.patient
            }
            Interaction.aux_date_created
            Interaction.aux_id
            Interaction.aux_clinical_data
            Interaction.given_name
            Interaction.family_name
            Interaction.gender
            Interaction.dob
            Interaction.city
            Interaction.phone_number
            Interaction.national_id
         }
         """;
   static final String INTERACTION_FIELD_NAMES =
         """
         uid
         Interaction.source_id {
            uid
            SourceId.facility
            SourceId.patient
         }
         Interaction.aux_date_created
         Interaction.aux_id
         Interaction.aux_clinical_data
         Interaction.given_name
         Interaction.family_name
         Interaction.gender
         Interaction.dob
         Interaction.city
         Interaction.phone_number
         Interaction.national_id
         """;

   static final String EXPANDED_INTERACTION_FIELD_NAMES =
         """
         uid
         Interaction.source_id {
            uid
            SourceId.facility
            SourceId.patient
         }
         Interaction.aux_date_created
         Interaction.aux_id
         Interaction.aux_clinical_data
         Interaction.given_name
         Interaction.family_name
         Interaction.gender
         Interaction.dob
         Interaction.city
         Interaction.phone_number
         Interaction.national_id
         ~GoldenRecord.interactions @facets(score) {
            uid
            GoldenRecord.source_id {
              uid
              SourceId.facility
              SourceId.patient
            }
            GoldenRecord.aux_date_created
            GoldenRecord.aux_auto_update_enabled
            GoldenRecord.aux_id
            GoldenRecord.given_name
            GoldenRecord.family_name
            GoldenRecord.gender
            GoldenRecord.dob
            GoldenRecord.city
            GoldenRecord.phone_number
            GoldenRecord.national_id
         }
         """;

   static final String QUERY_GET_INTERACTION_BY_UID =
         """
         query interactionByUid($uid: string) {
            all(func: uid($uid)) {
               uid
               Interaction.source_id {
                 uid
                 SourceId.facility
                 SourceId.patient
               }
               Interaction.aux_date_created
               Interaction.aux_id
               Interaction.aux_clinical_data
               Interaction.given_name
               Interaction.family_name
               Interaction.gender
               Interaction.dob
               Interaction.city
               Interaction.phone_number
               Interaction.national_id
            }
         }
         """;

   static final String QUERY_GET_GOLDEN_RECORD_BY_UID =
         """
         query goldenRecordByUid($uid: string) {
            all(func: uid($uid)) {
               uid
               GoldenRecord.source_id {
                  uid
                  SourceId.facility
                  SourceId.patient
               }
               GoldenRecord.aux_date_created
               GoldenRecord.aux_auto_update_enabled
               GoldenRecord.aux_id
               GoldenRecord.given_name
               GoldenRecord.family_name
               GoldenRecord.gender
               GoldenRecord.dob
               GoldenRecord.city
               GoldenRecord.phone_number
               GoldenRecord.national_id
            }
         }
         """;

   static final String QUERY_GET_EXPANDED_INTERACTIONS =
         """
         query expandedInteraction() {
            all(func: uid(%s)) {
               uid
               Interaction.source_id {
                  uid
                  SourceId.facility
                  SourceId.patient
               }
               Interaction.aux_date_created
               Interaction.aux_id
               Interaction.aux_clinical_data
               Interaction.given_name
               Interaction.family_name
               Interaction.gender
               Interaction.dob
               Interaction.city
               Interaction.phone_number
               Interaction.national_id
               ~GoldenRecord.interactions @facets(score) {
                  uid
                  GoldenRecord.source_id {
                    uid
                    SourceId.facility
                    SourceId.patient
                  }
                  GoldenRecord.aux_date_created
                  GoldenRecord.aux_auto_update_enabled
                  GoldenRecord.aux_id
                  GoldenRecord.given_name
                  GoldenRecord.family_name
                  GoldenRecord.gender
                  GoldenRecord.dob
                  GoldenRecord.city
                  GoldenRecord.phone_number
                  GoldenRecord.national_id
               }
            }
         }
         """;

   static final String QUERY_GET_GOLDEN_RECORDS =
         """
         query goldenRecord() {
            all(func: uid(%s)) {
               uid
               GoldenRecord.source_id {
                  uid
                  SourceId.facility
                  SourceId.patient
               }
               GoldenRecord.aux_date_created
               GoldenRecord.aux_auto_update_enabled
               GoldenRecord.aux_id
               GoldenRecord.given_name
               GoldenRecord.family_name
               GoldenRecord.gender
               GoldenRecord.dob
               GoldenRecord.city
               GoldenRecord.phone_number
               GoldenRecord.national_id
            }
         }
         """;

   static final String QUERY_GET_EXPANDED_GOLDEN_RECORDS =
         """
         query expandedGoldenRecord() {
            all(func: uid(%s), orderdesc: GoldenRecord.aux_date_created) {
               uid
               GoldenRecord.source_id {
                  uid
                  SourceId.facility
                  SourceId.patient
               }
               GoldenRecord.aux_date_created
               GoldenRecord.aux_auto_update_enabled
               GoldenRecord.aux_id
               GoldenRecord.given_name
               GoldenRecord.family_name
               GoldenRecord.gender
               GoldenRecord.dob
               GoldenRecord.city
               GoldenRecord.phone_number
               GoldenRecord.national_id
               GoldenRecord.interactions @facets(score) {
                  uid
                  Interaction.source_id {
                    uid
                    SourceId.facility
                    SourceId.patient
                  }
                  Interaction.aux_date_created
                  Interaction.aux_id
                  Interaction.aux_clinical_data
                  Interaction.given_name
                  Interaction.family_name
                  Interaction.gender
                  Interaction.dob
                  Interaction.city
                  Interaction.phone_number
                  Interaction.national_id
               }
            }
         }
         """;

   static final String MUTATION_CREATE_SOURCE_ID_TYPE =
         """
         type SourceId {
            SourceId.facility
            SourceId.patient
         }
         """;
       
   static final String MUTATION_CREATE_SOURCE_ID_FIELDS =
         """
         SourceId.facility:                     string    @index(exact)                      .
         SourceId.patient:                      string    @index(exact)                      .
         """;
         
   static final String MUTATION_CREATE_GOLDEN_RECORD_TYPE =
         """

         type GoldenRecord {
            GoldenRecord.source_id:                 [SourceId]
            GoldenRecord.aux_date_created
            GoldenRecord.aux_auto_update_enabled
            GoldenRecord.aux_id
            GoldenRecord.given_name
            GoldenRecord.family_name
            GoldenRecord.gender
            GoldenRecord.dob
            GoldenRecord.city
            GoldenRecord.phone_number
            GoldenRecord.national_id
            GoldenRecord.interactions:              [Interaction]
         }
         """;
           
   static final String MUTATION_CREATE_GOLDEN_RECORD_FIELDS =
         """
         GoldenRecord.source_id:                [uid]                                        .
         GoldenRecord.aux_date_created:         datetime                                     .
         GoldenRecord.aux_auto_update_enabled:  bool                                         .
         GoldenRecord.aux_id:                   string                                       .
         GoldenRecord.given_name:               string    @index(exact,trigram)              .
         GoldenRecord.family_name:              string    @index(exact,trigram)              .
         GoldenRecord.gender:                   string    @index(exact,trigram)              .
         GoldenRecord.dob:                      string                                       .
         GoldenRecord.city:                     string    @index(trigram)                    .
         GoldenRecord.phone_number:             string    @index(exact,trigram)              .
         GoldenRecord.national_id:              string    @index(exact,trigram)              .
         GoldenRecord.interactions:             [uid]     @reverse                           .
         """;

   static final String MUTATION_CREATE_INTERACTION_TYPE =
         """

         type Interaction {
            Interaction.source_id:                     SourceId
            Interaction.aux_date_created
            Interaction.aux_id
            Interaction.aux_clinical_data
            Interaction.given_name
            Interaction.family_name
            Interaction.gender
            Interaction.dob
            Interaction.city
            Interaction.phone_number
            Interaction.national_id
         }
         """;

   static final String MUTATION_CREATE_INTERACTION_FIELDS =
         """
         Interaction.source_id:                    uid                                          .
         Interaction.aux_date_created:             datetime                                     .
         Interaction.aux_id:                       string                                       .
         Interaction.aux_clinical_data:            string                                       .
         Interaction.given_name:                   string    @index(exact,trigram)              .
         Interaction.family_name:                  string    @index(exact,trigram)              .
         Interaction.gender:                       string                                       .
         Interaction.dob:                          string                                       .
         Interaction.city:                         string                                       .
         Interaction.phone_number:                 string                                       .
         Interaction.national_id:                  string    @index(exact,trigram)              .
         """;

   private CustomDgraphConstants() {}

}

File: JeMPI_Apps/JeMPI_Configuration/src/main/scala/configuration/CustomDgraphConstants.scala
package configuration

import java.io.{File, PrintWriter}

private object CustomDgraphConstants {

  private val classLocation =
    "../JeMPI_LibMPI/src/main/java/org/jembi/jempi/libmpi/dgraph"
  private val customClassName = "CustomDgraphConstants"
  private val packageText = "org.jembi.jempi.libmpi.dgraph"

  def generate(config: Config): Unit = {

    val classFile: String =
      classLocation + File.separator + customClassName + ".java"
    println("Creating " + classFile)
    val file: File = new File(classFile)
    val writer: PrintWriter = new PrintWriter(file)

    writer.println(s"""package $packageText;
         |
         |public final class $customClassName {
         |""".stripMargin)

    golden_record_predicates()
    interaction_predicates()
    golden_record_field_names()
    expanded_golden_record_field_names()
    interaction_field_names()
    expanded_interaction_field_names()
    query_get_interaction_by_uid()
    query_get_golden_record_by_uid()
    query_get_expanded_interactions()
    query_get_golden_records()
    query_get_expanded_golden_records()
    mutation_create_source_id_type()
    mutation_create_source_id_fields()
    mutation_create_golden_record_type()
    mutation_create_golden_record_fields()
    mutation_create_interaction_type()
    mutation_create_interaction_fields()

    writer.println(s"""   private $customClassName() {}
         |
         |}""".stripMargin)
    writer.flush()
    writer.close()

    def golden_record_predicates(): Unit =
      if (config.uniqueGoldenRecordFields.isDefined)
        config.uniqueGoldenRecordFields.get.zipWithIndex.foreach {
          case (field, _) =>
            val name = Utils.camelCaseToSnakeCase(field.fieldName)
            writer.println(
              s"""   public static final String PREDICATE_GOLDEN_RECORD_${name.toUpperCase} = "GoldenRecord.$name";""".stripMargin
            )
        }
      end if
      config.demographicFields.zipWithIndex.foreach { case (field, _) =>
        val name = Utils.camelCaseToSnakeCase(field.fieldName)
        writer.println(
          s"""   public static final String PREDICATE_GOLDEN_RECORD_${name.toUpperCase} = "GoldenRecord.$name";""".stripMargin
        )
      }
      writer.println(
        s"""   public static final String PREDICATE_GOLDEN_RECORD_INTERACTIONS = "GoldenRecord.interactions";""".stripMargin
      )
    end golden_record_predicates

    def interaction_predicates(): Unit =
      if (config.uniqueInteractionFields.isDefined)
        config.uniqueInteractionFields.get.zipWithIndex.foreach {
          case (field, _) =>
            val name = Utils.camelCaseToSnakeCase(field.fieldName)
            writer.println(
              s"""   public static final String PREDICATE_INTERACTION_${name.toUpperCase} = "Interaction.$name";""".stripMargin
            )
        }
      end if
      config.demographicFields.zipWithIndex.foreach { case (field, _) =>
        val fieldName = Utils.camelCaseToSnakeCase(field.fieldName)
        writer.println(
          s"""   public static final String PREDICATE_INTERACTION_${fieldName.toUpperCase} = "Interaction.$fieldName";""".stripMargin
        )
      }
    end interaction_predicates

    def golden_record_field_names(): Unit =
      writer.println(s"""
           |   static final String DEPRECATED_GOLDEN_RECORD_FIELD_NAMES =
           |         \"\"\"
           |         uid
           |         GoldenRecord.source_id {
           |            uid
           |            SourceId.facility
           |            SourceId.patient
           |         }""".stripMargin)
      if (config.uniqueGoldenRecordFields.isDefined)
        config.uniqueGoldenRecordFields.get.foreach(field => {
          val name = field.fieldName
          writer.println(s"         GoldenRecord.$name")
        })
      end if
      config.demographicFields.foreach(field => {
        val name = field.fieldName
        writer.println(s"         GoldenRecord.$name")
      })
      writer.println(s"""         \"\"\";""".stripMargin)
    end golden_record_field_names

    def expanded_golden_record_field_names(): Unit = {
      writer.println(s"""
           |   static final String DEPRECATED_EXPANDED_GOLDEN_RECORD_FIELD_NAMES =
           |         \"\"\"
           |         uid
           |         GoldenRecord.source_id {
           |            uid
           |            SourceId.facility
           |            SourceId.patient
           |         }""".stripMargin)

      if (config.uniqueGoldenRecordFields.isDefined) {
        config.uniqueGoldenRecordFields.get.foreach(field => {
          val name = field.fieldName
          writer.println(s"         GoldenRecord.$name")
        })
      }
      config.demographicFields.foreach(field => {
        val name = field.fieldName
        writer.println(s"         GoldenRecord.$name")
      })
      writer.println(s"""         GoldenRecord.interactions @facets(score) {
           |            uid
           |            Interaction.source_id {
           |               uid
           |               SourceId.facility
           |               SourceId.patient
           |            }""".stripMargin)
      if (config.uniqueInteractionFields.isDefined) {
        config.uniqueInteractionFields.get.foreach(field => {
          val name = field.fieldName
          writer.println(s"            Interaction.$name")
        })
      }
      config.demographicFields.foreach(field => {
        val name = field.fieldName
        writer.println(s"            Interaction.$name")
      })
      writer.println(s"""         }
           |         \"\"\";""".stripMargin)
    }

    def interaction_field_names(): Unit = {
      writer.println(
        s"""   static final String DEPRECATED_INTERACTION_FIELD_NAMES =
           |         \"\"\"
           |         uid
           |         Interaction.source_id {
           |            uid
           |            SourceId.facility
           |            SourceId.patient
           |         }""".stripMargin
      )

      if (config.uniqueInteractionFields.isDefined) {
        config.uniqueInteractionFields.get.foreach(field => {
          val name = field.fieldName
          writer.println(s"         Interaction.$name")
        })
      }
      config.demographicFields.foreach(field => {
        val name = field.fieldName
        writer.println(s"         Interaction.$name")
      })

      writer.println(s"""         \"\"\";
           |""".stripMargin)

    }

    def expanded_interaction_field_names(): Unit = {
      writer.println(
        s"""   static final String DEPRECATED_EXPANDED_INTERACTION_FIELD_NAMES =
           |         \"\"\"
           |         uid
           |         Interaction.source_id {
           |            uid
           |            SourceId.facility
           |            SourceId.patient
           |         }""".stripMargin
      )

      if (config.uniqueInteractionFields.isDefined) {
        config.uniqueInteractionFields.get.foreach(field => {
          val name = field.fieldName
          writer.println(s"         Interaction.$name")
        })
      }
      config.demographicFields.foreach(field => {
        val name = field.fieldName
        writer.println(s"         Interaction.$name")
      })

      writer.println(s"""         ~GoldenRecord.interactions @facets(score) {
           |            uid
           |            GoldenRecord.source_id {
           |              uid
           |              SourceId.facility
           |              SourceId.patient
           |            }""".stripMargin)

      if (config.uniqueGoldenRecordFields.isDefined) {
        config.uniqueGoldenRecordFields.get.foreach(field => {
          val name = field.fieldName
          writer.println(s"            GoldenRecord.$name")
        })
      }
      config.demographicFields.foreach(field => {
        val name = field.fieldName
        writer.println(s"            GoldenRecord.$name")
      })
      writer.println(s"""         }
           |         \"\"\";
           |""".stripMargin)

    }

    def query_get_interaction_by_uid(): Unit = {
      writer.println(
        s"""   static final String DEPRECATED_QUERY_GET_INTERACTION_BY_UID =
           |         \"\"\"
           |         query interactionByUid($$uid: string) {
           |            all(func: uid($$uid)) {
           |               uid
           |               Interaction.source_id {
           |                 uid
           |                 SourceId.facility
           |                 SourceId.patient
           |               }""".stripMargin
      )
      if (config.uniqueInteractionFields.isDefined) {
        config.uniqueInteractionFields.get.foreach(field => {
          val name = field.fieldName
          writer.println(s"               Interaction.$name")
        })
      }
      config.demographicFields.foreach(field => {
        val name = field.fieldName
        writer.println(s"               Interaction.$name")
      })
      writer.println(s"""            }
           |         }
           |         \"\"\";
           |""".stripMargin)
    }

    def query_get_golden_record_by_uid(): Unit = {
      writer.println(s"""   static final String QUERY_GET_GOLDEN_RECORD_BY_UID =
           |         \"\"\"
           |         query goldenRecordByUid($$uid: string) {
           |            all(func: uid($$uid)) {
           |               uid
           |               GoldenRecord.source_id {
           |                  uid
           |                  SourceId.facility
           |                  SourceId.patient
           |               }""".stripMargin)
      if (config.uniqueGoldenRecordFields.isDefined) {
        config.uniqueGoldenRecordFields.get.foreach(field => {
          val name = field.fieldName
          writer.println(s"               GoldenRecord.$name")
        })
      }
      config.demographicFields.foreach(field => {
        val name = field.fieldName
        writer.println(s"               GoldenRecord.$name")
      })
      writer.println(s"""            }
           |         }
           |         \"\"\";
           |""".stripMargin)
    }

    def query_get_expanded_interactions(): Unit = {
      writer.println(
        s"""   static final String QUERY_GET_EXPANDED_INTERACTIONS =
           |         \"\"\"
           |         query expandedInteraction() {
           |            all(func: uid(%s)) {
           |               uid
           |               Interaction.source_id {
           |                  uid
           |                  SourceId.facility
           |                  SourceId.patient
           |               }""".stripMargin
      )
      if (config.uniqueInteractionFields.isDefined) {
        config.uniqueInteractionFields.get.foreach(field => {
          val name = field.fieldName
          writer.println(s"               Interaction.$name")
        })
      }
      config.demographicFields.foreach(field => {
        val name = field.fieldName
        writer.println(s"               Interaction.$name")
      })
      writer.println(
        s"""               ~GoldenRecord.interactions @facets(score) {
           |                  uid
           |                  GoldenRecord.source_id {
           |                    uid
           |                    SourceId.facility
           |                    SourceId.patient
           |                  }""".stripMargin
      )
      if (config.uniqueGoldenRecordFields.isDefined) {
        config.uniqueGoldenRecordFields.get.foreach(field => {
          val name = field.fieldName
          writer.println(s"                  GoldenRecord.$name")
        })
      }
      config.demographicFields.foreach(field => {
        val name = field.fieldName
        writer.println(s"                  GoldenRecord.$name")
      })
      writer.println(s"""               }
           |            }
           |         }
           |         \"\"\";
           |""".stripMargin)
    }

    def query_get_golden_records(): Unit = {
      writer.println(s"""   static final String QUERY_GET_GOLDEN_RECORDS =
           |         \"\"\"
           |         query goldenRecord() {
           |            all(func: uid(%s)) {
           |               uid
           |               GoldenRecord.source_id {
           |                  uid
           |                  SourceId.facility
           |                  SourceId.patient
           |               }""".stripMargin)
      if (config.uniqueGoldenRecordFields.isDefined) {
        config.uniqueGoldenRecordFields.get.foreach(field => {
          val name = field.fieldName
          writer.println(s"               GoldenRecord.$name")
        })
      }
      config.demographicFields.foreach(field => {
        val name = field.fieldName
        writer.println(s"               GoldenRecord.$name")
      })
      writer.println(s"""            }
           |         }
           |         \"\"\";
           |""".stripMargin)
    }

    def query_get_expanded_golden_records(): Unit = {
      writer.println(s"""   static final String QUERY_GET_EXPANDED_GOLDEN_RECORDS =
           |         \"\"\"
           |         query expandedGoldenRecord() {
           |            all(func: uid(%s), orderdesc: GoldenRecord.aux_date_created) {
           |               uid
           |               GoldenRecord.source_id {
           |                  uid
           |                  SourceId.facility
           |                  SourceId.patient
           |               }""".stripMargin)
      if (config.uniqueGoldenRecordFields.isDefined) {
        config.uniqueGoldenRecordFields.get.foreach(field => {
          val name = field.fieldName
          writer.println(s"               GoldenRecord.$name")
        })
      }
      config.demographicFields.foreach(field => {
        val name = field.fieldName
        writer.println(s"               GoldenRecord.$name")
      })
      writer.println(
        s"""               GoldenRecord.interactions @facets(score) {
           |                  uid
           |                  Interaction.source_id {
           |                    uid
           |                    SourceId.facility
           |                    SourceId.patient
           |                  }""".stripMargin
      )
      if (config.uniqueInteractionFields.isDefined) {
        config.uniqueInteractionFields.get.foreach(field => {
          val name = field.fieldName
          writer.println(s"                  Interaction.$name")
        })
      }
      config.demographicFields.foreach(field => {
        val name = field.fieldName
        writer.println(s"                  Interaction.$name")
      })
      writer.println(s"""               }
           |            }
           |         }
           |         \"\"\";
           |""".stripMargin)
    }

    def mutation_create_source_id_type(): Unit = {
      writer.println(
        s"""   static final String DEPRECATED_MUTATION_CREATE_SOURCE_ID_TYPE =
           |         \"\"\"
           |         type SourceId {
           |            SourceId.facility
           |            SourceId.patient
           |         }
           |         \"\"\";
       """.stripMargin
      )
    }

    def mutation_create_source_id_fields(): Unit = {
      writer.println(s"""   static final String DEPRECATED_MUTATION_CREATE_SOURCE_ID_FIELDS =
           |         \"\"\"
           |         SourceId.facility:                     string    @index(exact)                      .
           |         SourceId.patient:                      string    @index(exact)                      .
           |         \"\"\";
         """.stripMargin)
    }

    def mutation_create_golden_record_type(): Unit = {
      writer.println(
        s"""   static final String DEPRECATED_MUTATION_CREATE_GOLDEN_RECORD_TYPE =
           |         \"\"\"
           |
           |         type GoldenRecord {
           |            GoldenRecord.source_id:                 [SourceId]""".stripMargin
      )
      if (config.uniqueGoldenRecordFields.isDefined) {
        config.uniqueGoldenRecordFields.get.foreach(field => {
          val name = field.fieldName
          writer.println(s"            GoldenRecord.$name")
        })
      }
      config.demographicFields.foreach(field => {
        val name = field.fieldName
        writer.println(s"            GoldenRecord.$name")
      })
      writer.println(
        s"""            GoldenRecord.interactions:              [Interaction]
           |         }
           |         \"\"\";
           """.stripMargin
      )
    }

    def mutation_create_golden_record_fields(): Unit = {
      writer.println(
        s"""   static final String DEPRECATED_MUTATION_CREATE_GOLDEN_RECORD_FIELDS =
           |         \"\"\"
           |         GoldenRecord.source_id:                [uid]     @reverse                           .""".stripMargin
      )
      if (config.uniqueGoldenRecordFields.isDefined) {
        config.uniqueGoldenRecordFields.get.foreach(field => {
          val name = field.fieldName
          val index = field.index.getOrElse("")
          val fieldType =
            (if field.isList.isDefined && field.isList.get then "["
             else "") + field.fieldType.toLowerCase + (if field.isList.isDefined && field.isList.get
                                                       then "]"
                                                       else "")
          writer.println(
            s"""${" " * 9}GoldenRecord.$name:${" " * (25 - name.length)}$fieldType${" " * (10 - fieldType.length)}$index${" " * (35 - index.length)}.""".stripMargin
          )
        })
      }
      config.demographicFields
        .foreach(field => {
          val name = field.fieldName
          val index = field.indexGoldenRecord.getOrElse("")
          val fieldType =
            (if field.isList.isDefined && field.isList.get then "["
             else "") + field.fieldType.toLowerCase + (if field.isList.isDefined && field.isList.get
                                                       then "]"
                                                       else "")
          writer.println(
            s"""${" " * 9}GoldenRecord.$name:${" " * (25 - name.length)}$fieldType${" " * (10 - fieldType.length)}$index${" " * (35 - index.length)}.""".stripMargin
          )
        })
      writer.println(
        s"""         GoldenRecord.interactions:             [uid]     @reverse                           .
           |         \"\"\";
           |""".stripMargin
      )
    }

    def mutation_create_interaction_type(): Unit =
      writer.println(
        s"""   static final String DEPRECATED_MUTATION_CREATE_INTERACTION_TYPE =
           |         \"\"\"
           |
           |         type Interaction {
           |            Interaction.source_id:                     SourceId""".stripMargin
      )
      if (config.uniqueInteractionFields.isDefined) {
        config.uniqueInteractionFields.get.foreach(field => {
          val name = field.fieldName
          writer.println(s"            Interaction.$name")
        })
      }
      config.demographicFields.foreach(field => {
        val name = field.fieldName
        writer.println(s"            Interaction.$name")
      })
      writer.println(s"""         }
           |         \"\"\";
           |""".stripMargin)
    end mutation_create_interaction_type

    def mutation_create_interaction_fields(): Unit = {
      writer.println(
        s"""   static final String DEPRECATED_MUTATION_CREATE_INTERACTION_FIELDS =
           |         \"\"\"
           |         Interaction.source_id:                    uid                                          .""".stripMargin
      )
      if (config.uniqueInteractionFields.isDefined) {
        config.uniqueInteractionFields.get.foreach(field => {
          val name = field.fieldName
          val index = field.index.getOrElse("")
          val fieldType =
            (if field.isList.isDefined && field.isList.get then "["
             else "") + field.fieldType.toLowerCase + (if field.isList.isDefined && field.isList.get
                                                       then "]"
                                                       else "")
          writer.println(
            s"""${" " * 9}Interaction.$name:${" " * (29 - name.length)}$fieldType${" " * (10 - fieldType.length)}$index${" " * (35 - index.length)}.""".stripMargin
          )
        })
      }
      config.demographicFields.foreach(field => {
        val name = field.fieldName
        val index = field.indexInteraction.getOrElse("")
        val fieldType = field.fieldType.toLowerCase
        writer.println(
          s"""${" " * 9}Interaction.$name:${" " * (29 - name.length)}$fieldType${" " * (10 - fieldType.length)}$index${" " * (35 - index.length)}.""".stripMargin
        )
      })
      writer.println(s"""         \"\"\";
           |""".stripMargin)
    }

  }
  end generate

}
